app/core/ingestion/physical_document.py

```diff
@@ -32,21 +36,53 @@ def create_physical_document_from_params(
     db.flush()
     result["physical_document"] = to_dict(physical_document)
 
-    for language in params.language:
+    update_physical_document_languages(db, params.language, result, physical_document)
+
+    return physical_document
+
+
+def update_physical_document_languages(
+    db: Session,
+    langs: list[str],
+    result: dict[str, Any],
+    physical_document: PhysicalDocument,
+) -> None:
+    """
+    Updates the physical document with the languages in param.
+
+    :param Session db: connection to the db.
+    :param list[str] langs: List of languages.
+    :param dict[str, Any] result: The result of the ingest
+    :param PhysicalDocument physical_document: The physical document to update
+    """
+    existing_lang_ids = [
+        pdl.language_id
+        for pdl in db.query(PhysicalDocumentLanguage)
+        .filter_by(document_id=physical_document.id)
+        .all()
+    ]
+
+    for language in langs:
         lang = db.query(Language).filter(Language.name == language).one_or_none()
-        if lang is not None:
-            doc_languages = result.get("language", [])
-            doc_languages.append(to_dict(lang))
-            result["language"] = doc_languages
+        if lang is None:
+            continue
 
-            physical_document_language = PhysicalDocumentLanguage(
-                language_id=lang.id, document_id=physical_document.id
-            )
-            db.add(physical_document_language)
-            db.flush()
+        if lang.id in existing_lang_ids:
+            continue
 
-            phys_doc_languages = result.get("physical_document_language", [])
-            phys_doc_languages.append(to_dict(physical_document_language))
-            result["physical_document_language"] = phys_doc_languages
+        doc_languages = result.get("language", [])
+        doc_languages.append(to_dict(lang))
+        result["language"] = doc_languages
 
-    return physical_document
+        physical_document_language = PhysicalDocumentLanguage(
+            language_id=lang.id,
+            document_id=physical_document.id,
+            source=LanguageSource.USER,
```

Comment:
I'm just wondering whether this function would be utilised by the db writer when calling back to the backend with identified languages and thus whether defaulting to a USER defined language is correct.

Reply:
Nope this is ingestion only - i.e. via CSV ... i.e. via a user :smile: